bionic/aip/future.py

```diff
@@ -0,0 +1,96 @@
+import blocks
+import logging
+import time
+import googleapiclient.discovery
+from concurrent.futures import Future as _Future, TimeoutError, CancelledError
+from enum import Enum, auto
+
+
+class AIPError(Exception):
+    pass
+
+
+class State(Enum):
+    STATE_UNSPECIFIED = auto()
+    QUEUED = auto()
+    PREPARING = auto()
+    RUNNING = auto()
+    SUCCEEDED = auto()
+    FAILED = auto()
+    CANCELLING = auto()
+    CANCELLED = auto()
+
+
+EXECUTE = {State.STATE_UNSPECIFIED, State.QUEUED, State.PREPARING, State.RUNNING}
+CANCEL = {State.CANCELLING, State.CANCELLED}
+FAIL = {State.FAILED}
+
+
+class Future(_Future):
+    def __init__(self, project: str, job_id: str, output: str):
+        self.project = project
+        self.job_id = job_id
+        self.output = output
+        self.aip = googleapiclient.discovery.build("ml", "v1", cache_discovery=False)
+
+    def name(self):
+        return f"projects/{self.project}/jobs/{self.job_id}"
+
+    def cancel(self):
+        request = self.aip.projects().jobs().cancel(name=self.name())
+        request.execute()
+        return True
+
+    def query(self):
+        request = self.aip.projects().jobs().get(name=self.name())
+        response = request.execute()
+        return State[response["state"]], response.get("errorMessage", "")
+
+    def cancelled(self):
+        # TODO do we ever want to distinguish between canceling and cancelled on GCP?
+        # at the moment we just trust that GCP will succesfully cancel once requested
+        state, _ = self.query()
+        return state in CANCEL
+
+    def running(self):
+        state, _ = self.query()
+        return state in EXECUTE
+
+    def done(self):
+        state, _ = self.query()
+        return state is State.SUCCEEDED or state in CANCEL
+
+    def result(self, timeout: int = None):
+        # TODO support other serializers
+        exc = self.exception(timeout)
+        if exc is not None:
+            raise exc
+
+        try:
+            return blocks.unpickle(self.output)
+        except ValueError:
+            logging.warning(f"Failed to load output from succesful job at {self.path}")
+            raise
+
+    def exception(self, timeout: int = None):
```

Comment:
I know what you mean. This is a weird interface, more so if you also consider that it sometimes returns errors and sometimes raises them. But this is the [concurrent.futures.Future](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future) interface we tried to follow to be consistent with parallel execution.

Reply:
Oh, I didn't realize that! That makes more sense, sorry.

Reply:
Agree i would not write it this way in a vacuum, but the consistency still seems good in the long run

Reply:
I would also have initially agreed and tried to do it this way, but after seeing the interaction with `wait`, I'm starting to think that `concurrent.futures.Future` is not intended to be subclassed. If we want to mix AIP jobs with our existing multiprocessing futures (and `wait` on all of them together), it seems like we don't have much choice but to wrap this inside a "real" future -- backed by a separate thread handling the polling logic -- rather than exposing this interface directly.

That said, I don't think we need to change this in this PR.

Reply:
Yes that makes sense, let me simplify this down to what we will actually use and then I will leave the wrapping in a multiprocessing future up to a subsequent PR.